server_interfaces_test.go

```diff
@@ -0,0 +1,348 @@
+	tester, err := newTester()
+	if err != nil {
+		t.Errorf("Unexpected error: %s", err)
+	}
+	conn, err := SessionBus()
+	if err != nil {
+		t.Errorf("Unexpected error: %s", err)
+	}
+	obj := conn.Object(tester.Name(), "/com/github/godbus/tester")
+	var out string
+	err = obj.Call("com.github.godbus.dbus.Tester.Test", 0).Store(&out)
+	if err == nil {
+		t.Errorf("didn't get expected error")
+	}
+	tester.Close()
+}
+
+func TestHandlerInvalidArgType(t *testing.T) {
+	tester, err := newTester()
+	if err != nil {
+		t.Errorf("Unexpected error: %s", err)
+	}
+	conn, err := SessionBus()
+	if err != nil {
+		t.Errorf("Unexpected error: %s", err)
+	}
+	obj := conn.Object(tester.Name(), "/com/github/godbus/tester")
+	var out string
+	err = obj.Call("com.github.godbus.dbus.Tester.Test", 0, 2.10).Store(&out)
+	if err == nil {
+		t.Errorf("didn't get expected error")
+	}
+	tester.Close()
+}
+
+func TestHandlerIntrospect(t *testing.T) {
+	tester, err := newTester()
+	if err != nil {
+		t.Errorf("Unexpected error: %s", err)
+	}
+	conn, err := SessionBus()
+	if err != nil {
+		t.Errorf("Unexpected error: %s", err)
+	}
+	obj := conn.Object(tester.Name(), "/com/github/godbus/tester")
+	var out string
+	err = obj.Call("org.freedesktop.DBus.Introspectable.Introspect", 0).Store(&out)
+	expected := `<node>
+    <interface name="org.freedesktop.DBus.Introspectable.Introspect">
+        <method name="Introspect">
+            <arg name="out" type="i" direction="out">
+        </method>
+    </interface>
+    <interface name="com.github.godbus.dbus.Tester">
+        <method name="Test">
+            <arg name="in" type="i" direction="in">
+            <arg name="out" type="i" direction="out">
+        </method>
+        <signal name="sig1">
+            <arg name="out" type="i" direction="out">
+        </signal>
+    </interface>
+</node>`
+	if out != expected {
+		t.Errorf("didn't get expected return value, expected %s got %s", expected, out)
+	}
+	tester.Close()
+}
+
+func TestHandlerIntrospectPath(t *testing.T) {
+	tester, err := newTester()
+	if err != nil {
+		t.Errorf("Unexpected error: %s", err)
+	}
+	conn, err := SessionBus()
+	if err != nil {
+		t.Errorf("Unexpected error: %s", err)
+	}
+	obj := conn.Object(tester.Name(), "/com/github/godbus")
+	var out string
+	err = obj.Call("org.freedesktop.DBus.Introspectable.Introspect", 0).Store(&out)
+	expected := `<node><node name="tester"></node></node>`
+	if out != expected {
+		t.Errorf("didn't get expected return value, expected %s got %s", expected, out)
+	}
+	tester.Close()
+}
+
+func TestHandlerSignal(t *testing.T) {
+	tester, err := newTester()
+	if err != nil {
+		t.Errorf("Unexpected error: %s", err)
+	}
+	conn, err := SessionBus()
+	if err != nil {
+		t.Errorf("Unexpected error: %s", err)
+	}
+	tester.conn.BusObject().(*Object).AddMatchSignal("com.github.godbus.dbus.Tester", "sig1")
+	conn.Emit("/com/github/godbus/tester",
+		"com.github.godbus.dbus.Tester.sig1", "foo")
+	sig := <-tester.sigs
```

Comment:
If a signal isn't sent then this just hangs forever. You should use a timeout to fail the test.